src/realm/sync/noinst/pending_bootstrap_store.cpp

```diff
@@ -0,0 +1,293 @@
+        cur_changeset.set(m_changeset_origin_timestamp, int64_t(changesets[idx].origin_timestamp));
+        cur_changeset.set(m_changeset_original_changeset_size, int64_t(changesets[idx].original_changeset_size));
+        BinaryData compressed_data(compressed_changesets[idx].data(), compressed_changesets[idx].size());
+        cur_changeset.set(m_changeset_data, compressed_data);
+    }
+
+    tr->commit();
+
+    if (did_create) {
+        m_logger->trace("Created new pending bootstrap object for query version %1", query_version);
+    }
+    else {
+        m_logger->trace("Added batch to pending bootstrap object for query version %1", query_version);
+    }
+    if (progress) {
+        m_logger->trace("Finalized pending bootstrap object for query version %1", query_version);
+    }
+    m_has_pending = true;
+}
+
+bool PendingBootstrapStore::has_pending()
+{
+    return m_has_pending;
+}
+
+void PendingBootstrapStore::clear()
+{
+    auto tr = m_db->start_write();
+    auto bootstrap_table = tr->get_table(m_table);
+    bootstrap_table->clear();
+    tr->commit();
+}
+
+PendingBootstrapStore::PendingBatch PendingBootstrapStore::peek_pending(size_t limit_in_bytes)
+{
+    auto tr = m_db->start_read();
+    auto bootstrap_table = tr->get_table(m_table);
+    if (bootstrap_table->is_empty()) {
+        return {};
+    }
+
+    // We should only have one pending bootstrap at a time.
+    REALM_ASSERT(bootstrap_table->size() == 1);
+
+    auto bootstrap_obj = bootstrap_table->get_object(0);
+    PendingBatch ret;
+    ret.query_version = bootstrap_obj.get<int64_t>(m_query_version);
+
+    if (!bootstrap_obj.is_null(m_progress)) {
+        auto progress_obj = bootstrap_obj.get_linked_object(m_progress);
+        SyncProgress progress;
+        progress.latest_server_version.version = progress_obj.get<int64_t>(m_progress_latest_server_version);
+        progress.latest_server_version.salt = progress_obj.get<int64_t>(m_progress_latest_server_version_salt);
+        progress.download.server_version = progress_obj.get<int64_t>(m_progress_download_server_version);
+        progress.download.last_integrated_client_version =
+            progress_obj.get<int64_t>(m_progress_download_client_version);
+        progress.upload.last_integrated_server_version = progress_obj.get<int64_t>(m_progress_upload_server_version);
+        progress.upload.client_version = progress_obj.get<int64_t>(m_progress_upload_client_version);
+        ret.progress = std::move(progress);
+    }
+
+    auto changeset_list = bootstrap_obj.get_linklist(m_changesets);
+    size_t bytes_so_far = 0;
+    for (size_t idx = 0; idx < changeset_list.size() && bytes_so_far < limit_in_bytes; ++idx) {
+        auto cur_changeset = changeset_list.get_object(idx);
+        ret.changeset_data.push_back(util::AppendBuffer<char>());
+        auto& uncompressed_buffer = ret.changeset_data.back();
+
+        auto raw_changeset_data = cur_changeset.get<BinaryData>(m_changeset_data);
```

Comment:
nit: you could name this compressed_changeset_data

Reply:
done